File: individual-name-matching/node-examples/append-simkeys-to-file.js
// Simple Node.js example that:
// - Reads full names from a text file (one per line)
// - Calls Interzoid's Full Name Match API for each name
// - Writes the original value and returned SimKey to a CSV file
//
// Requirements:
// - Node.js 18+ (for built-in fetch)
// - No third-party libraries

const fs = require("fs");
const readline = require("readline");

// 1. Replace this with your API key from https://www.interzoid.com/manage-api-account
const API_KEY = "YOUR_API_KEY_HERE";

// 2. Hardcoded input and output file names
const inputFileName = "sample-input-file.txt";
const outputFileName = "output.csv";

// Helper function to escape a value for CSV:
// - Wraps in double quotes if needed
// - Escapes existing double quotes by doubling them
function csvEscape(value) {
  const str = String(value);
  if (/[",\r\n]/.test(str)) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

async function main() {
  // Open the input file as a readable stream
  let inputStream;
  try {
    inputStream = fs.createReadStream(inputFileName, { encoding: "utf8" });
  } catch (err) {
    console.error("Error opening input file:", err.message);
    return;
  }

  // Create the output CSV file as a writable stream
  let outputStream;
  try {
    outputStream = fs.createWriteStream(outputFileName, { encoding: "utf8" });
  } catch (err) {
    console.error("Error creating output file:", err.message);
    return;
  }

  // Use readline to read the input file line by line
  const rl = readline.createInterface({
    input: inputStream,
    crlfDelay: Infinity,
  });

  let lineNumber = 0;

  for await (const line of rl) {
    lineNumber++;
    const originalValue = line;

    // Skip blank lines (optional)
    if (originalValue.length === 0) {
      continue;
    }

    // Build the API URL with proper URL-encoding.
    // URLSearchParams will take care of encoding spaces and special characters.
    const endpoint = "https://api.interzoid.com/getfullnamematch";

    const params = new URLSearchParams({
      license: API_KEY,
      fullname: originalValue,
    });

    const apiURL = `${endpoint}?${params.toString()}`;

    try:
      // Make the HTTP GET request
      const response = await fetch(apiURL);

      // If the response is not OK (e.g. 4xx or 5xx), log an error and write an empty SimKey
      if (!response.ok) {
        console.error(
          `API error on line ${lineNumber} (${JSON.stringify(
            originalValue
          )}): HTTP ${response.status} ${response.statusText}`
        );
        outputStream.write(
          `${csvEscape(originalValue)},${csvEscape("")}\n`
        );
        continue;
      }

      // Read and parse the JSON response
      const data = await response.json();

      // Write one CSV row per input line: original value and SimKey
      const simKey = data.SimKey || "";
      outputStream.write(
        `${csvEscape(originalValue)},${csvEscape(simKey)}\n`
      );
    } catch (err) {
      // If there is a network or parsing error, log it and still write a row
      console.error(
        `Error for line ${lineNumber} (${JSON.stringify(
          originalValue
        )}):`,
        err.message
      );
      outputStream.write(
        `${csvEscape(originalValue)},${csvEscape("")}\n`
      );
    }
  }

  // Close the output file when done
  outputStream.end(() => {
    console.log("Done. Results written to", outputFileName);
  });
}

// Run the script
main().catch((err) => {
  console.error("Unexpected error:", err.message);
});

File: company-name-matching/java-examples/append-simkeys-to-file.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.HttpURLConnection;
import java.net.URLEncoder;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * Reads company names from a text file, calls Interzoid's company match API
 * for each line, and writes the original value + SimKey to a CSV file.
 *
 * File name: append-simkeys-to-file.java
 * Run with:  java AppendSimkeysToFile
 */
class AppendSimkeysToFile {

    // Replace this with your API key from https://www.interzoid.com/manage-api-account
    private static final String API_KEY = "YOUR_API_KEY_HERE";

    // Hardcoded input and output file names (same as Go example)
    private static final String INPUT_FILE_NAME = "sample-input-file.txt";
    private static final String OUTPUT_FILE_NAME = "output.csv";

    // Simple POJO to hold parsed JSON response
    private static class ResponseData {
        String simKey;
        String code;
        String credits;
    }

    public static void main(String[] args) {
        try (
            BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE_NAME));
            PrintWriter writer = new PrintWriter(new BufferedWriter(new FileWriter(OUTPUT_FILE_NAME)))
        ) {
            String line;
            int lineNumber = 0;

            while ((line = reader.readLine()) != null) {
                lineNumber++;
                String originalValue = line;

                // Skip completely empty lines (same as Go example)
                if (originalValue.length() == 0) {
                    continue;
                }

                ResponseData result;
                try {
                    result = callApiForCompany(originalValue);
                } catch (Exception e) {
                    System.out.printf("Error calling API for line %d (%q): %s%n",
                            lineNumber, originalValue, e.getMessage());
                    // On error, write original value and empty SimKey
                    writer.println(toCsvRow(originalValue, ""));
                    continue;
                }

                if (result == null) {
                    System.out.printf("No response for line %d (%q)%n", lineNumber, originalValue);
                    writer.println(toCsvRow(originalValue, ""));
                    continue;
                }

                // Optional: check result.code for "Success"
                if (!"Success".equalsIgnoreCase(result.code)) {
                    System.out.printf(
                        "Non-success code for line %d (%q): Code=%s%n",
                        lineNumber, originalValue, result.code
                    );
                }

                // Write original value and SimKey as a CSV row
                writer.println(toCsvRow(originalValue, result.simKey));
            }

            System.out.println("Done. Results written to " + OUTPUT_FILE_NAME);

        } catch (IOException e) {
            System.out.println("Error opening/creating files: " + e.getMessage());
        }
    }

    /**
     * Calls the Interzoid getcompanymatchadvanced API for a given company name.
     */
    private static ResponseData callApiForCompany(String companyName) throws IOException {
        String companyParam = URLEncoder.encode(companyName, StandardCharsets.UTF_8.toString());

        String apiUrl =
            "https://api.interzoid.com/getcompanymatchadvanced?license=" + API_KEY +
            "&company=" + companyParam +
            "&algorithm=model-v4-wide";

        URL url = new URL(apiUrl);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");

        int status = conn.getResponseCode();
        BufferedReader in;
        if (status >= 200 && status < 300) {
            in = new BufferedReader(
                    new java.io.InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        } else {
            in = new BufferedReader(
                    new java.io.InputStreamReader(conn.getErrorStream(), StandardCharsets.UTF_8));
        }

        StringBuilder responseBuilder = new StringBuilder();
        String inputLine;
        while ((inputLine = in.readLine()) != null) {
            responseBuilder.append(inputLine);
        }
        in.close();
        conn.disconnect();

        String json = responseBuilder.toString().trim();
        if (json.isEmpty()) {
            return null;
        }

        ResponseData data = new ResponseData();
        data.simKey = extractJsonValue(json, "SimKey");
        data.code = extractJsonValue(json, "Code");
        data.credits = extractJsonValue(json, "Credits");
        return data;
    }

    /**
     * Very simple JSON value extractor for flat JSON like:
     * {"SimKey":"...","Code":"Success","Credits":"1"}
     *
     * This avoids external JSON libraries.
     */
    private static String extractJsonValue(String json, String key) {
        String quotedKey = "\"" + key + "\"";
        int keyIndex = json.indexOf(quotedKey);
        if (keyIndex == -1) {
            return "";
        }

        int colonIndex = json.indexOf(':', keyIndex + quotedKey.length());
        if (colonIndex == -1) {
            return "";
        }

        // Skip whitespace
        int valueStart = colonIndex + 1;
        while (valueStart < json.length() &&
               Character.isWhitespace(json.charAt(valueStart))) {
            valueStart++;
        }

        // Handle quoted string value
        if (valueStart < json.length() && json.charAt(valueStart) == '"') {
            valueStart++;
            StringBuilder sb = new StringBuilder();
            boolean escaping = false;

            for (int i = valueStart; i < json.length(); i++) {
                char c = json.charAt(i);
                if (escaping) {
                    sb.append(c);
                    escaping = false;
                } else if (c == '\\') {
                    escaping = true;
                } else if (c == '"') {
                    break;
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        } else {
            // Fallback for non-string values (not expected here)
            int valueEnd = valueStart;
            while (valueEnd < json.length() &&
                   json.charAt(valueEnd) != ',' &&
                   json.charAt(valueEnd) != '}' &&
                   !Character.isWhitespace(json.charAt(valueEnd))) {
                valueEnd++;
            }
            return json.substring(valueStart, valueEnd).trim();
        }
    }

    /**
     * Escapes a single field for CSV (RFC-4180-ish):
     * - Wrap in quotes if it contains comma, quote, or newline
     * - Double any internal quotes
     */
    private static String csvEscape(String field) {
        if (field == null) {
            field = "";
        }
        boolean mustQuote = field.contains(",") ||
                            field.contains("\"") ||
                            field.contains("\n") ||
                            field.contains("\r");
        if (!mustQuote) {
            return field;
        }
        String escaped = field.replace("\"", "\"\"");
        return "\"" + escaped + "\"";
    }

    /**
     * Builds a single CSV row with two columns.
     */
    private static String toCsvRow(String col1, String col2) {
        return csvEscape(col1) + "," + csvEscape(col2);
    }
}
